runtime/Cpp/runtime/src/Recognizer.h

```diff
@@ -17,35 +18,38 @@ namespace antlr4 {
 
     Recognizer();
     Recognizer(Recognizer const&) = delete;
-    virtual ~Recognizer();
+    virtual ~Recognizer() = default;
 
     Recognizer& operator=(Recognizer const&) = delete;
 
-    virtual std::vector<std::string> const& getRuleNames() const = 0;
+    virtual antlrcpp::Span<const std::string_view> getRuleNames() const = 0;
 
     /**
      * Get the vocabulary used by the recognizer.
      *
      * @return A {@link Vocabulary} instance providing information about the
      * vocabulary used by the grammar.
      */
-    virtual dfa::Vocabulary const& getVocabulary() const = 0;
+    virtual const Vocabulary& getVocabulary() const = 0;
 
     /// <summary>
     /// Get a map from token names to token types.
     /// <p/>
     /// Used for XPath and tree pattern compilation.
     /// </summary>
-    virtual std::map<std::string_view, size_t> getTokenTypeMap();
+    const std::unordered_map<std::string_view, size_t>& getTokenTypeMap() const;
```

Comment:
I'd probably have all these map types typedefs in the public section of the class, so that it is easier to change and avoids repetition of the same otherwise somewhat lengthy type.